hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferContainer.java

```diff
@@ -0,0 +1,345 @@
+                    + " is too large. The current tailSequence is:" + tailSequence);
+        }
+
+        if (readSequence < headSequence) {
+            throw new StaleSequenceException("sequence:" + readSequence
+                    + " is too small. The current headSequence is:" + headSequence, headSequence);
+        }
+    }
+
+    public long add(Data item) {
+        return addInternal(item);
+    }
+
+    private long addInternal(Data dataItem) {
+        tailSequence++;
+
+        if (tailSequence - capacity == headSequence) {
+            headSequence++;
+        }
+
+        int index = toIndex(tailSequence);
+
+        Object item = dataItem;
+        if (inMemoryFormat == OBJECT) {
+            item = serializationService.toObject(dataItem);
+        }
+
+        // first we write the dataItem in the ring.
+        ringItems[index] = item;
+
+        // and then we optionally write the expiration.
+        if (isRetentionEnabled()) {
+            ringExpirationMs[index] = currentTimeMillis() + retentionMillis;
+        }
+
+        return tailSequence;
+    }
+
+    public long addAll(Data[] items) {
+        long result = -1;
+        for (Data item : items) {
+            result = addInternal(item);
+        }
+        return result;
+    }
+
+    public Data read(long sequence) {
+        checkReadSequence(sequence);
+
+        int index = toIndex(sequence);
+        Object item = ringItems[index];
+        return serializationService.toData(item);
+    }
+
+    /**
+     * @param beginSequence the sequence of the first item to read.
+     * @param result        the List where the result are stored in.
+     * @return returns the sequenceId of the next item to read. This is needed if not all required items are found.
+     */
+    public long readMany(long beginSequence, ReadResultSetImpl result) {
+        checkReadSequence(beginSequence);
+
+        long seq = beginSequence;
+        while (seq <= tailSequence) {
+            int index = toIndex(seq);
+            Object item = ringItems[index];
+
+            result.addItem(item);
+
+            seq++;
+
+            if (result.isMaxSizeReached()) {
+                // we have found all items we are looking for. We are done.
+                break;
+            }
+        }
+
+        return seq;
+    }
+
+    /**
+     * Cleans up the ringbuffer by deleting all expired items.
+     */
+    public void cleanup() {
+        if (!isRetentionEnabled() || headSequence > tailSequence) {
+            return;
+        }
+
+        long now = currentTimeMillis();
+        while (headSequence <= tailSequence) {
+            int index = toIndex(headSequence);
+
+            if (ringExpirationMs[index] > now) {
+                return;
+            }
+
+            //System.out.println("expiring index:" + index);
+
+            // we null the slot and allow the gc to take care of the object.
+            ringItems[index] = null;
```

Comment:
shouldn't we clean the `ringExpirationMs[index]` too ? 
I understand that it may not be a problem since you are incrementing headSequence but mentioning just for completeness